README.md
# **Advanced Source Package Managment (spkg)**
![spkg - v1.6.0](https://img.shields.io/badge/spkg-1.6.0-success) 
![Code Size](https://img.shields.io/github/languages/code-size/Juliandev02/spkg)
![Commit activity](https://img.shields.io/github/commit-activity/w/Juliandev02/spkg)
![License](https://img.shields.io/github/license/Juliandev02/spkg)

spkg is a package manager that downloads the source code from the official sources, and then mostly compiles it specifically for your device. 

The goal of spkg is to get the latest versions of programs easily and 
without much experience, even under distros that do not offer the latest version.
By compiling the package, the program is optimized for your device and can run faster.
So spkg offers you a high security, so you don't have to worry about viruses in packages.

spkg also offers pre-build packages, so you don't have to build big packages like Java yourself. 
<br>

# **Plugins**
spkg offers a convenient plugin system, allowing users to enable advanced features such as installing packages over a secure sandbox! 

## **Sandbox  ![spkg-sandbox](https://img.shields.io/badge/spkg--sandbox-important)**

![spkg-sandbox](https://img.shields.io/badge/spkg--sandbox-1.3.0-informational)
![sandbox-release](https://img.shields.io/badge/Release-stable-success)
![sandbox-methods](https://img.shields.io/badge/Methods-bwrap%2C%20chroot-important)
![sandbox-architetures](https://img.shields.io/badge/Architectures-x86__64%2C%20x86%2C%20aarch64-success)

spkg-sandbox installs packages in a separate environment from the system to provide more security and compatibility. spkg-sandbox works on any system, regardless of the package manager, while spkg without spkg-sandbox currently only works on deb and apk-based systems. It uses Ubuntu 22.04 or 20.04 as base. 
<br>

## **Docker Containers  ![spkg-docker](https://img.shields.io/badge/spkg--docker-important)**

![spkg-docker](https://img.shields.io/badge/spkg--docker-0.1.0-informational)
![docker-release](https://img.shields.io/badge/Release-Not%20released-critical)
![docker-methods](https://img.shields.io/badge/Methods-docker%2C%20podman%20(%3F)-important)
![docker-architetures](https://img.shields.io/badge/Architectures-depending%20on%20docker%20image-success)

spkg-docker installs packages in a Docker container. This provides more security, and is more convenient. Therefore, spkg works on any system that supports docker. The Docker image used is an (custom) Ubuntu 22.04 or 20.04 image.
<br><br>

## **Supported Platforms**
spkg was built on the basis of Python 3.9. We recommend to use Python 3.10 or higher. All Python versions below 3.9 are untested and therefore not recommended. Therefore, please do not open an issue if spkg does not work under Python 3.8 or lower. 

| Version       | Supported | Status                                                    |
| --            | --        | --                                                        | 
| Python 3.8    | ?         | ![](https://img.shields.io/badge/-Untested-orange)        |
| Python 3.9    | Yes       | ![](https://img.shields.io/badge/-Tested,_Working-success)|
| Python 3.10   | Yes       | ![](https://img.shields.io/badge/-Tested,_Working-success)|
| Python 3.11   | Yes       | ![](https://img.shields.io/badge/-Tested,_Working-success)|
| Python 3.12   | ?         | ![](https://img.shields.io/badge/-Untested-orange)        |

## **Linux Support**
Distrobutions and versions that are not listed are not supported by us. 
| Platform  | Version       | Supported     | Branch                                                    | Package Method    |
| --        | --            | --            | --                                                        | --                |   
| Debian    | 9 and below   | No            | ![](https://img.shields.io/badge/-End_of_Life-red)        | -
| Debian    | 10            | Yes           | ![](https://img.shields.io/badge/-OldStable-yellowgreen)  | native, sandbox, docker
| Debian    | 11            | Yes           | ![](https://img.shields.io/badge/-Stable-success)         | native, sandbox, docker
| Debian    | 12            | Yes           | ![](https://img.shields.io/badge/-Future-success)  | native, sandbox, docker
| Debian    | Sid           | Yes           | ![](https://img.shields.io/badge/-Unstable-orange)        | native, sandbox, docker
| Ubuntu    | 16.04         | No            | ![](https://img.shields.io/badge/-ESM-orange)             | -
| Ubuntu    | 18.04         | Not offical   | ![](https://img.shields.io/badge/-LTS-yellowgreen)        | sandbox, docker
| Ubuntu    | 20.04         | Yes           | ![](https://img.shields.io/badge/-LTS-green)              | native, sandbox, docker
| Ubuntu    | 22.04         | Yes           | ![](https://img.shields.io/badge/-LTS-success)            | native, sandbox, docker
| Ubuntu    | 22.10         | Yes           | ![](https://img.shields.io/badge/-Old_Stable-green)       | native, sandbox, docker
| Ubuntu    | 23.04         | Yes           | ![](https://img.shields.io/badge/-Stable-success)         | native, sandbox, docker
| Alpine    | 3.16.x        | Untested      | ![](https://img.shields.io/badge/-Stable-green)           | native, sandbox, docker
| Alpine    | 3.17.x        | Yes           | ![](https://img.shields.io/badge/-Stable-green)           | native, sandbox, docker
| Alpine    | 3.18.x        | Yes           | ![](https://img.shields.io/badge/-Stable-success)         | native, sandbox, docker
| Fedora    | 37 - 38       | Yes           | ![](https://img.shields.io/badge/-Stable-success)         | (soon native), sandbox, docker
| Arch      | -             | Yes           | ![](https://img.shields.io/badge/-Bleeding%20Edge-blueviolet) | sandbox, docker
| Gentoo    | -             | Untested      | ![](https://img.shields.io/badge/-RR-informational)       | sandbox, docker


CHANGELOG.txt
* -- spkg v1.6.0 -- *
- (Important) We're changing our package build system to our new PKGBUILD-System (spkg-pbs)
- (Important) New Codebase, Data Structure, Package- and Repository Structure. Older Versions of spkg are now discontinued and no longer work
- (Important) spkg uses now a new directory for the databases. (/var/lib/spkg/)
- (New) New release variables
- (New) Complete new modular-system for spkg (defs-system)
- (New) Completly new package build scripts
- (New) spkg lists now the packages in alphabetic order
- (New) New Database Sync system with multiple repository support
- (New) Check Mark when something passed (e.g. package db synced successfully)
- (New) New Text-Displaying for the install function
- (New) Updated the spkg installer for use with the new package build system
- (Package-related) Support for drivers added (Beta)
- (Package-related) Update for package groups
- (Package-related) Package Subdirectories/Categories, e.g. drivers/rtw89
- (Package-related) 5 main package branches, main, community, unstable and git 
- (Package-related) Ported almost all package to the new build system
- (Package-related) Remove function for the python*-packages (currently untested/beta)
- (Package-related) New package cdn (core-cdn)
- (Package-related) Removed the spkg-bin Package because it's slower and unstable
- (Optimization) Much more code comments and documentation 
- (Optimization) Much more modular code 
- (Optimization) Python Object Classes for many functions
- (Optimization) CLI-Design update
- (Optimization) Added build date for *Release Types* "rc", "beta" and "alpha"
- (Optimization) Print a warning if user configuration folder doesn't exist
- (Fix) Better formatting
- (Fix) Fixed bug not showing package information when arch equals all
- (Fix) Fixed a translation bug for the Missing Permissons String in the english language
- (Fix) Fixed a typo for "Synchronisiert" 
- (Fix) Fixed now showing "Installing some base packages" in sandbox plugin when debug mode is not enabled
- (Backend) Many fixes and better handling for almost everything
- (Backend) Flag for selecting using the ./data as the data directory (Development only!)
- (Backend) Moved plugin_daemon.py to the src Directory
- (Backend) Optimization of the error in case of an incorrect language configuration
- (Backend) Repomgr Update (1.0.1)
- (Backend) Sandbox Update (1.3.0) (Fixes, New Design, Comments, etc)
- (Code) Added License information header to every source files

* -- Known Issues -- *
- (Issue) -

sandbox.py
"""
Sandbox System for spkg - Made by Juliandev02
"""

import json
from sqlite3 import *
from colorama import Fore
from sys import exit
import os
import platform
import time

# Define Home Directory
home_dir = os.getenv("HOME")

# Use User Home even if spkg was executed with sudo
if os.environ.get('SUDO_USER'):
    home_dir = os.path.expanduser(f"~{os.environ['SUDO_USER']}")
else:
    home_dir = os.path.expanduser("~")

# Language Config
spkg_config = "/etc/spkg/config.json"
with open(spkg_config, "r") as f:
    spkg_cfg = json.load(f)

language = spkg_cfg['language']

if not language == "de" and not language == "en":
    exit()
    
# Basic Variables
bootstrap_location = f"{home_dir}/.local/spkg/sandbox/"
dist = "jammy"

BOLD = '\033[1m'
UNDERLINE = '\033[4m'
RESET = '\033[0m'


with open('/etc/os-release') as f:
    os_info = dict(line.strip().split('=') for line in f if '=' in line)

# Language Strings
if language == "de":
    Description = "spkg-sandbox installiert Pakete in einer isolierten Umgebung."

elif language == "en":
    Description = "spkg-sandbox installs packages in a isolated environment."

# Spec Class for more Details about the Plugin
class Spec:
    Name = "spkg-sandbox"
    Desc = Description
    Version = "0.2.1"
    Commands = f"""
    -> setup
    -> config
    -> remove
    -> enter
    """

# PluginHandler Main Class
class PluginHandler:
    def setup():
        print(f"{Fore.YELLOW + BOLD}[!]{Fore.RESET + RESET} Detecting Operating System")
        print(f"{Fore.GREEN + BOLD}[!]{Fore.RESET + RESET} Detected {os_info['NAME']}")
        
        if not os.path.exists("/usr/sbin/debootstrap"):
            print(f"{Fore.RED + BOLD}Error:{Fore.RESET + RESET} spkg-sandbox cannot be executed on your system. Missing dependency 'debootstrap'")
            exit()
        
        if os_info['ID'] == "debian" and os_info['VERSION_ID'] == '"10"':
            print(f"{Fore.YELLOW + BOLD}Warning:{Fore.RESET + RESET} Your version of debootstrap is outdated and doesn't support to build Ubuntu 22.04 Jammy Jellyfish.")
            print(f"         Using Focal Fossa (Ubuntu 20.04) build script instead ...")
            dist = "focal"
            
        elif os_info['ID'] == "debian" and os_info['VERSION_ID'] == '"11"':
            print(f"{Fore.YELLOW + BOLD}Warning:{Fore.RESET + RESET} Your version of debootstrap is outdated and doesn't support to build Ubuntu 22.04 Jammy Jellyfish.")
            print(f"         Using Focal Fossa (Ubuntu 20.04) build script instead ...")
            dist = "focal"
        
        elif os_info['ID'] == "ubuntu" and os_info['VERSION_ID'] == '"20.04"':
            print(f"{Fore.YELLOW + BOLD}Warning:{Fore.RESET + RESET} Your version of ubuntu is outdated and the sandbox cannot continue to work with Ubuntu 22.04 Jammy Jellyfish")
            print(f"         Using Focal Fossa (Ubuntu 20.04) build script instead ...")
            dist = "focal"
        
        else:
            print(f"{Fore.YELLOW + BOLD}Warning:{Fore.RESET + RESET} Your Linux distrobution has not yet been tested by the spkg developers. It is possible that spkg-sandbox does not work. Please open a GitHub issue if something is not working. ")
            
        print(f"{Fore.YELLOW + BOLD}[!]{Fore.RESET + RESET} Sandbox Setup will now start")
        
        try: 
            ans = input("Do you want to continue? [Y/N]? ")
            
        except KeyboardInterrupt: 
            print("\nAborting ...")
            exit()
        
        if ans != "y" and ans != "Y" and ans != "j" and ans != "J": 
            print("Aborting ...")
            exit()
            
            
        print(f"{Fore.YELLOW + BOLD}[!]{Fore.RESET + RESET} Checking system architecture")
        arch = platform.machine()
        
        if arch == "x86_64":
            arch = "amd64"
            repo = "http://archive.ubuntu.com/ubuntu"
            
        elif arch == "x86":
            arch = "i386"
            repo = "http://archive.ubuntu.com/ubuntu"
            
        elif arch == "aarch64":
            arch = "arm64"
            repo="http://ports.ubuntu.com/ubuntu-ports"
        
        else:
            print(f"{Fore.RED + BOLD}Error:{Fore.RESET + RESET} spkg-sandbox cannot be executed on your system. Your architecture is currently not supported.")
            exit()
            
        if os.path.exists(bootstrap_location):
            print(f"{Fore.YELLOW + BOLD}Warning:{Fore.RESET + RESET} You have already a installation of spkg-sandbox. Reinstalling spkg-sandbox ...")
            os.system(f"sudo rm -rf {bootstrap_location}")
        
        start_time = time.time()
        
        print(f"{Fore.YELLOW + BOLD}[!]{Fore.RESET + RESET} Bootstrapping your spkg-sandbox ... This could take some time depending on your drive speed and internet speed")
        os.mkdir(bootstrap_location)
        os.system(f"sudo debootstrap --arch={arch} --variant=minbase --include=systemd,libsystemd0,wget,ca-certificates,busybox-static {dist} {bootstrap_location} {repo}")
        
        print(f"{Fore.YELLOW + BOLD}[!]{Fore.RESET + RESET} Updating your sandbox ...")
        os.system(f"sudo chroot {bootstrap_location} apt clean all &> /dev/null && apt autoclean &> /dev/null && apt update &> /dev/null")
        
        print(f"{Fore.YELLOW + BOLD}[!]{Fore.RESET + RESET} Modifying /etc/apt/sources.list for the best program compability")
        os.system(f"sudo rm {bootstrap_location}/etc/apt/sources.list")
        
        if arch == "x86_64" or arch == "x86":
            os.system(f'sudo tee -a {bootstrap_location}/etc/apt/sources.list <<< "deb http://archive.ubuntu.com/ubuntu {dist} main restricted universe multiverse"')
            os.system(f'sudo tee -a {bootstrap_location}/etc/apt/sources.list <<< "deb http://archive.ubuntu.com/ubuntu {dist}-backports main restricted universe multiverse"')
            os.system(f'sudo tee -a {bootstrap_location}/etc/apt/sources.list <<< "deb http://archive.ubuntu.com/ubuntu {dist}-proposed main restricted universe multiverse"')
            os.system(f'sudo tee -a {bootstrap_location}/etc/apt/sources.list <<< "deb http://archive.ubuntu.com/ubuntu {dist}-security main restricted universe multiverse"')
            os.system(f'sudo tee -a {bootstrap_location}/etc/apt/sources.list <<< "deb http://archive.ubuntu.com/ubuntu {dist}-updates main restricted universe multiverse"')
            
        else:   
            os.system(f'sudo tee -a {bootstrap_location}/etc/apt/sources.list <<< "deb http://ports.ubuntu.com/ubuntu-ports {dist} main restricted universe multiverse"')
            os.system(f'sudo tee -a {bootstrap_location}/etc/apt/sources.list <<< "deb http://ports.ubuntu.com/ubuntu-ports {dist}-backports main restricted universe multiverse"')
            os.system(f'sudo tee -a {bootstrap_location}/etc/apt/sources.list <<< "deb http://ports.ubuntu.com/ubuntu-ports {dist}-proposed main restricted universe multiverse"')
            os.system(f'sudo tee -a {bootstrap_location}/etc/apt/sources.list <<< "deb http://ports.ubuntu.com/ubuntu-ports {dist}-security main restricted universe multiverse"')
            os.system(f'sudo tee -a {bootstrap_location}/etc/apt/sources.list <<< "deb http://ports.ubuntu.com/ubuntu-ports {dist}-updates main restricted universe multiverse"')  
            
        print(f"{Fore.YELLOW + BOLD}[!]{Fore.RESET + RESET} Installing some base packages ... ")
        os.system(f"sudo chroot {bootstrap_location} apt update &> /dev/null")
        os.system(f"sudo chroot {bootstrap_location} apt install -y python3 python3-dev python3-pip #&> /dev/null")
        
        end_time = time.time()
        
        print(f"{Fore.GREEN + BOLD}[!]{Fore.RESET + RESET} Finished sandbox setup in {round(start_time - end_time, 2)} s")
        
    def config():
        print("config")
        
    def remove():
        print(f"{Fore.YELLOW + BOLD}[!]{Fore.RESET + RESET}Removing sandbox ... This can take some time.")
        os.system(f"sudo rm -rf {bootstrap_location}")
        exit()
        
    def enter():
        os.system(f'sudo chroot {bootstrap_location}')